jarbas_hive_mind/message.py

```diff
@@ -0,0 +1,143 @@
+    # from mycroft-bus in slave
+    BROADCAST = "broadcast"  # forward message to all slaves
+    PROPAGATE = "propagate"  # forward message to all slaves and masters
+    ESCALATE = "escalate"  # forward message up the authority chain to all
+    # masters
+    HELLO = "hello"  # like escalate, used to announce the device
+    QUERY = "query"  # like escalate, but stops once one of the nodes can
+    # send a response
+    CASCADE = "cascade"  # like propagate, but expects a response back from
+    # all nodes in the hive (responses optional)
+    PING = "ping"  # like cascade, but used to map the network
+    RENDEZVOUS = "rendezvous"  # reserved for rendezvous-nodes
+
+    THIRDPRTY = "3rdparty"  # user land message, do whatever you want
+
+
+class HiveMessage:
+    def __init__(self, msg_type, payload=None, node=None, source_peer=None,
+                 route=None, target_peers=None):
+        #  except for the hivemind node classes receiving the message and
+        #  creating the object nothing should be able to change these values
+        #  node classes might change them a runtime by the private attribute
+        #  but end-users should consider them read_only
+        if msg_type not in [m.value for m in HiveMessageType]:
+            raise ValueError("Unknown HiveMessage.msg_type")
+
+        self._msg_type = msg_type
+        # the payload is more or less a free for all
+        # the msg_type determines what happens to the message, but the
+        # payload can simply be ignored by the receiving module
+
+        # some msg_types might return HiveMessage, others (mycroft) Message
+        # we should support the dict/json format, json is used at the
+        # transport layer before converting into any of these formats
+        if isinstance(payload, Message):
+            payload = {"type": payload.msg_type,
+                       "data": payload.data,
+                       "context": payload.context}
+        elif isinstance(payload, str):
+            payload = json.loads(payload)
+        self._payload = payload or {}
+
+        self._node = node  # node semi-unique identifier
+        self._source_peer = source_peer  # peer_id
+        self._route = route or []  # where did this message come from
+        self._targets = target_peers or []  # where will it be sent
+
+    @property
+    def msg_type(self):
+        return self._msg_type
+
+    @property
+    def node_id(self):
+        return self._node
+
+    @property
+    def source_peer(self):
+        return self._source_peer
+
+    @property
+    def target_peers(self):
+        if self.source_peer:
+            return self._targets or [self._source_peer]
+        return self._targets
+
+    @property
+    def route(self):
+        return [r for r in self._route if r.get("targets") and r.get("source")]
+
+    @property
+    def payload(self):
+        if self.msg_type in [HiveMessageType.BUS, HiveMessageType.SHARED_BUS]:
```

Comment:
I'm not actually sure, but I think this would be more performant with either a generator or `is`/`or`, simply because this way it has to construct the list, then iterate over it